nbsite/nbbuild.py

```diff
@@ -178,34 +187,170 @@ class FixNotebookLinks(Preprocessor):
+        nb_filepath:
+            Location of the notebook which source to find, relative to rootdir.
+
+        Returns
+        -------
+        source_relpath: str | None
+            The source file path relative to root notebook. If the source file
+            is not found, issues a warning and returns none.
+        """
+        for source_relpath in cls._iter_source_file_candidates(nb_filepath):
+            target_abspath = os.path.normpath(os.path.join(rootdir, source_relpath))
+            if cls.file_exists(target_abspath):
+                return source_relpath
+
+    @staticmethod
+    def _extract_links(markdown_text: str) -> Iterable[Tuple[str, str]]:
+        """Extract links to Notebook files (.ipynb) from markdown text and
+        yield them. Returns the full markdown link and the link target
+        separately.
+
+        Examples
+        --------
+        Markdown link: "[a](../foo/b.ipynb#spam)"
+        Link target: "..foo/b.ipynb"
+
+        String "foo [a](b.ipynb) bar [c](..foo/d.ipynb#spam)" would return
+        iterable, which when converted to a list would be: [
+             [('[a](b.ipynb)', 'b.ipynb')]
+             [('[c](../foo/d.ipynb#spam)', '../foo/d.ipynb')]
+        ]
+
+        """
+        for match in re.finditer(r"(\[.+?\]\((.+?\.ipynb)#*?.*?\))", markdown_text):
+            yield match.groups()
+
+    @classmethod
+    def _iter_source_file_candidates(cls, nb_filepath: str) -> Iterable[str]:
+        """Gets potential link targets corresponding to a notebook file path.
+        Potential link targets (source files paths) are formed by using the
+        relative path of the link target notebook, and adding the file
+        extensions from `cls.file_types`. In addition if the the filename part
+        of the `nb_filepath` starts with digits + one of ('-', '_', ' '),
+        the stem without the number prefix is used to form a potential link
+        target.
+
+        Parameters
+        ----------
+        nb_filepath:
+            The filepath of a link target a notebook. Must end with ".ipynb".
+            The links are relative to the directory of the notebook which
+            contains the markdown link. Example: "../foo/10-Some.ipynb"
+
+        Example
+        -------
+        >>> list(_iter_source_file_candidates("../foo/01-Some_Notebook.ipynb"))
+        [
+            "../foo/01-Some_Notebook.rst",
+            "../foo/01-Some_Notebook.md",
+            "../foo/Some_Notebook.rst",
+            "../foo/Some_Notebook.md",
+        ],
+        """
+
+        if not nb_filepath.endswith('.ipynb'):
+            raise ValueError('nb_filepath must end with .ipynb')
+
+        nb_path_without_extension = nb_filepath[:-6]
+        for extension in cls.file_types:
+            yield f"{nb_path_without_extension}.{extension}"
+
+        for extension in cls.file_types:
+            *directory, stem = re.split('(/)', nb_path_without_extension)
```

Comment:
I think you can use `os.path.split` which will simplify things.

Reply:
That's a good idea! It simplifies the code a bit. Also not sure if this is supposed to support builds on Windows which has different path separator 🤔.  Changes done.